onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h

```diff
@@ -302,11 +286,27 @@ class GQAAttentionBase {
             }
           }
 
-          // Calculate softcap / softmax
           if (softcap_ > 0.f) {
             ComputeAttentionSoftcapInplace(output_softmax + start_offset, static_cast<int>(window_size),
                                            static_cast<U>(softcap_));
           }
+
+          // Apply custom attention mask if there is any
+          // TODO (#23982): Implement masking during softmax computation in GQA CPU operator
+          if (attention_mask_thread != nullptr) {
+            if constexpr (std::is_same_v<U, T>) {
+              ApplyAttentionMask(output_softmax + start_offset, attention_mask_thread + start_offset,
+                                 static_cast<int>(window_size));
+            } else {
+              size_t bytes = window_size * sizeof(float);
```

Comment:
Add a check of U and T here like
static_assert(std::is_same_v<U, float> && std::is_same_v<T, MLFloat16>);